Update test logic to simulate NodeReady/False and NodeReady/Unknown events correctly

Huang-Wei opened the pull request:
**What type of PR is this?**

/kind failing-test

**What this PR does / why we need it**:

In 3 subtests of this test, we applied `NodeReady/False` for a particular Node and wait for the taint for 30 seconds. However, when node lifecycle manager gets that event, it pushes the Node object into a tainting queue:

```
2016:I1106 17:13:25.959358  105933 node_lifecycle_controller.go:834] Node node-1 is NotReady as of 2019-11-06 17:13:25.959339722 +0000 UTC m=+262.274940037. Adding it to the Taint queue.
```

Later on, the tainting worker pulls the Node object from the queue, it fetches the latest event - in other words, it might not be the original `NodeReady/False` event we expected, in the flake test, it's `NodeReady/Unknown` instead of `NodeReady/False`.

```
2027:I1106 17:13:25.973676  105933 controller_utils.go:193] Added [&Taint{Key:node.kubernetes.io/unreachable,Value:,Effect:NoExecute,TimeAdded:2019-11-06 17:13:25.96340613 +0000 UTC m=+262.279006529,}] Taint to Node node-1
```

It's because internally if a Node doesn't report heartbeat within a certain period (5 seconds?), node lifecycle manager will treat it as Unknown, which semantically indicates the Node is losing network connection.

Although the 2 conditions (False and Unknow) both implies a Node as unavailable, they're literally different types of conditions. So in our test, we expect the taint to be "NotReady:NoExecute", but it turns out to be "Unreachable:NoExecute", so the test sometimes fails.

**Which issue(s) this PR fixes**:

May fix #83321, in particular, https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/84738/pull-kubernetes-integration/1192120553314455554/.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig scheduling
/cc @ahg-g @ravisantoshgudimetla @damemi
/priority important-soon

Huang-Wei commented:
/retest

liggitt commented:
/retest

Huang-Wei commented:
/retest

ahg-g commented:
/retest

damemi commented on test/integration/scheduler/taint_test.go:
```diff
@@ -787,18 +790,47 @@ func TestTaintBasedEvictions(t *testing.T) {
+				} else if c == v1.ConditionUnknown {
+					// If we try to update the node with condition NotReady/Unknown, i.e. expect a Unreachable:NoExecute taint,
+					// we need to only send the update event once to simulate the network unreachable scenario.
 					heartbeatChans[i] <- struct{}{}
```
I might just be missing it, but do we still send a heartbeat to the other (notneeded) nodes? From what I can tell this loop [skips all the others](https://github.com/kubernetes/kubernetes/pull/84883/files#diff-d51d948272758524cc6289bc6f0b8daaR795-R797) so I can't find where the heartbeat is sent to those as well, though I could be wrong

Huang-Wei replied:
> do we still send a heartbeat to the other (notneeded) nodes?

Yes, in the beginning, when we instantiate the 3 goroutines, they're all sending NodeReady/True to APIServer: 


```go
			// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
			var heartbeatChans []chan struct{}
			for i := 0; i < nodeCount; i++ {
				heartbeatChans = append(heartbeatChans, make(chan struct{}))
			}
			for i := 0; i < nodeCount; i++ {
				// Spin up <nodeCount> goroutines to send heartbeat event to APIServer periodically.
				go func(i int) {
					for {
						select {
						case <-heartbeatChans[i]:
							return
						case <-time.Tick(heartbeatInternal):
							nodes[i].Status.Conditions = []v1.NodeCondition{
								{
									Type:              v1.NodeReady,
									Status:            v1.ConditionTrue,
									LastHeartbeatTime: metav1.Now(),
								},
							}
							updateNodeStatus(cs, nodes[i])
						}
					}
				}(i)
			}
```

And in this loop, if current node is "neededNode" we intentionally do (1) keep sending NodeReady/False events, or (2) only send NodeReady/Unknow once.

damemi replied:
sgtm, so the call to `updateNodeStatus` with now as the `lastHeartbeatTime` serves that purpose? I was looking for another spot where we actually passed something to the heartbeat channel, so I think I misunderstood that

damemi commented:
@Huang-Wei thanks for finding this and for the very detailed and helpful explanation. I had one question about the heartbeats but otherwise lgtm

ahg-g commented:
> if a Node doesn't report heartbeat
I am probably missing something, but the test is setup such that the nodes do report heartbeat, so is that happening?

ahg-g commented:
> > if a Node doesn't report heartbeat
> 
> I am probably missing something, but the test is setup such that the nodes do report heartbeat, so is that happening?

ok I get it now, for the "needed node", we don't send heartbeats, and so there is a race condition between the test setting the condition and the controller declaring the condition unknown

ahg-g commented on test/integration/scheduler/taint_test.go:
```diff
@@ -746,7 +749,7 @@ func TestTaintBasedEvictions(t *testing.T) {
+						case <-time.Tick(heartbeatInternal):
 							nodes[i].Status.Conditions = []v1.NodeCondition{
 								{
 									Type:              v1.NodeReady,
```
instead of the logic you added below, can we just update the neededNode status here consistently?:

```
c := []v1.NodeCondition{{Type: v1.NodeReady, Status: v1.ConditionTrue, LastHeartbeatTime: metav1.Now(),},
 if nodes[i] == neededNode {
   c = test.nodeConditions
}
nodes[i].Status.Conditions = c
...
```

Huang-Wei replied:
Sounds good, let me update.

Huang-Wei commented:
> ok I get it now, for the "needed node", we don't send heartbeats, and so there is a race condition between the test setting the condition and the controller declaring the condition unknown

Correct. We should precisely control the heartbeat, in one of the following 3 cases:

- keep sending NodeReady:True
- keep sending NodeReady:False (network is still healthy)
- sending NodeReady:Unknown once (network is unhealthy)

Huang-Wei commented:
@ahg-g @damemi Did some optimizations:

- optimize code to use one loop to spin up goroutines
- add `defer cleanupTest()` to avoid goroutine leaks
- use only one heartbeat channel

Please take another look.

Huang-Wei committed:
Update test logic to simulate NodeReady/False and NodeReady/Unknown events correctly

- optimize code to use one loop to spin up goroutines
- add `defer cleanupTest()` to avoid goroutine leaks
- use only one heartbeat channel

ahg-g commented on test/integration/scheduler/taint_test.go:
```diff
@@ -786,18 +763,55 @@ func TestTaintBasedEvictions(t *testing.T) {
+						// If it's expected to update the node with condition NotReady/Unknown,
+						// i.e. expect a Unreachable:NoExecute taint,
+						// we need to only send the update event once to simulate the network unreachable scenario.
+						nodeCopy := nodeCopyWithConditions(nodes[i], test.nodeConditions)
```
what happens if you keep updating the status with NotReady/Unknown? if the status sticks, then we don't need to special case it here, which should simplify the code alot.

Huang-Wei replied:
Technically, we can, but I'm not sure I like it, even I was thinking about not sending the Unknown event at all. It's because we started the node lifecycle manager, so we should let it handle the "network lost" case totally.

ahg-g commented on test/integration/scheduler/taint_test.go:
```diff
@@ -786,18 +763,55 @@ func TestTaintBasedEvictions(t *testing.T) {
 
+			// Regularly send heartbeat event to APIServer so that the cluster doesn't enter fullyDisruption mode.
+			// TODO(Huang-Wei): use "NodeDisruptionExclusion" feature to simply the below logic when it's beta.
+			heartbeatChan := make(chan struct{})
```
we don't really need the heartbeatChan at all, the context should suffice since it is created per test.

Huang-Wei replied:
not really, if we reuse the context, before closing the context (called in `defer cleanTest()`), the nodes have been cleaned up (in `cleanNodes()`), so there would still be a couple of attempts updating the deleted node, then it fails the test.

ahg-g commented on test/integration/scheduler/taint_test.go:
```diff
@@ -665,6 +667,7 @@ func TestTaintBasedEvictions(t *testing.T) {
 			podTolerations.SetExternalKubeInformerFactory(externalInformers)
 
 			context = initTestScheduler(t, context, true, nil)
+			defer cleanupTest(t, context)
```
we should call this for the other test above (TestTaintBasedEvictions).

Huang-Wei replied:
You mean `TestTaintNodeByCondition` above?

ahg-g replied:
yes, sorry for the confusion.

Huang-Wei replied:
it's now covered in https://github.com/kubernetes/kubernetes/commit/cd675cca922585677d58413fae93296ea307a916

ahg-g commented:
/lgtm

ahg-g commented:
/hold

ahg-g commented:
/lgtm
/hold cancel

ahg-g commented:
/retest
